Introduce tiled mesh index support (v2)

nyalldawson opened the pull request:
This is an alternative approach to https://github.com/qgis/QGIS/pull/53950 which addresses the review concerns

nyalldawson committed:
Introduce QgsTiledMeshTile

Represents an individual tile from a tiled mesh scene

nyalldawson committed:
Introduce QgsTiledMeshIndex

This is an equivalent of the point cloud index class, with some
notable differences:

- The class is designed to be thread safe. There's a shallow copy
QgsTiledMeshIndex class, which contains an implicitly shared
QgsAbstractTiledMeshIndex object. Tiled mesh data providers will
be accompanied by a concrete class of QgsAbstractTiledMeshIndex.
- The QgsTiledMeshIndex class takes care of thread safety by
protecting the underlying QgsAbstractTiledMeshIndex via a read/write
lock
- Callers can request tiled mesh tiles from the index by calling
QgsTiledMeshIndex::getTiles along with a QgsTiledMeshRequest object.
This will return the unique IDs of tiles matching the request.
Currently only geometric error based filtering is supported
by QgsTiledMeshRequest, but bounding box based filtering will
also be introduced.
- Individual tile details are retrieved by calling
QgsTiledMeshIndex::getTile along with the tile's unique ID.
- A caller can determine whether a specific tile can be further
refined by checking QgsTiledMeshIndex::tileCanRefine. This allows
the caller to check whether more children are available for a
tile, but which haven't yet been fetched (eg due to a sub dataset
tileset which hasn't yet been retrieved)
- A tile can be refined by calling QgsTiledMeshIndex::refineAsync.
(Non async refinement will be introduced later)
- Clients can request tile content by calling
  QgsTiledMeshIndex::retrieveContent. This will either return
a cached version of the content (if available), or retrieve and cache
the content for future retrieval. Currently this uses a custom caching
mechanism, but this will be replaced in future with
QgsTileDownloadManager. (Currently QgsTileDownloadManager lacks support
for blocking gets which is required by the index to avoid use
of local event loops)

The api is implemented by the cesium tiles data provider, which
builds the index upfront from the cesium tiles tileset.json definitions.

nyalldawson committed:
Add method to transform a tiled mesh bounding volume via QgsMatrix4x4

nyalldawson committed:
Add QgsTiledMeshRequest

Encapsulates parameters for filtering tiled mesh tiles

nyalldawson committed:
Add class for storing hierarchy tree of QgsTiledMeshTile

wonder-sk commented on src/core/tiledmesh/qgscesiumtilesdataprovider.cpp:
```diff
@@ -28,26 +28,430 @@
+  const QString contentUri = it.value()->tile()->resources().value( QStringLiteral( "content" ) ).toString();
+  locker.unlock();
+
+  return contentUri.endsWith( QLatin1String( ".json" ), Qt::CaseInsensitive );
```
The spec is a bit confusing about this.

https://github.com/CesiumGS/3d-tiles/tree/main/specification#tile-json says:

"A file extension is not required for content.uri. A content’s [tile format](https://github.com/CesiumGS/3d-tiles/tree/main/specification#core-tile-format-specifications) can be identified by the magic field in its header, or else as an external tileset if the content is JSON."

https://github.com/CesiumGS/3d-tiles/tree/main/specification#file-extensions-and-media-types also says that `.json` should be used, only to add at the end that "Explicit file extensions are optional. Valid implementations may ignore it and identify a content’s format by the magic field in its header."

So I wonder if we can just stick to `.json` test or if we should be checking magic field (which would be more complicated)

nyalldawson replied:
That's rather... annoying :upside_down_face: . 

I've implemented a work around for this. If there's an extension which itself reveals definitely whether or not the content can be a tileset then we use that information. Otherwise we pessimistically report that the tile has remote children available and handle the json/non-json detection when trying to fetch the children.

wonder-sk commented on src/core/tiledmesh/qgstiledmeshindex.h:
```diff
@@ -0,0 +1,239 @@
+     *
+     * \see tileCanRefine()
+     */
+    virtual bool refineAsync( const QString &id, QgsFeedback *feedback = nullptr ) = 0;
```
I am a bit confused about the naming - with `refineAsync` I thought it would be non-blocking... yet it blocks.

nyalldawson replied:
Fixed

wonder-sk commented on src/core/tiledmesh/qgstiledmeshindex.h:
```diff
@@ -0,0 +1,239 @@
+     * \see tileCanRefine()
+     */
+    virtual bool refineAsync( const QString &id, QgsFeedback *feedback = nullptr ) = 0;
+
```
One more note about naming - the term "refine" is being used in two slightly different scenarios:
- tile refinement strategy (add / replace)
- refine tile (to load next bit of hierarchy)

Maybe the second "refine" could be changed to something else? (in point clouds I think we used the term to "fetch hierarchy")

nyalldawson replied:
Ok, I've gone with fetchHierarchy and `Qgis::TileChildrenAvailability childAvailability( const QString &id )` instead.

wonder-sk commented on src/core/tiledmesh/qgscesiumtilesdataprovider.cpp:
```diff
@@ -28,26 +28,430 @@
+  return QStringList();
+}
+
+QStringList QgsCesiumTiledMeshIndex::getTiles( const QgsTiledMeshRequest &request, const QString &parentId )
```
getTiles() implementation will get the tileCanRefine() + refineAsync() support later I guess?

nyalldawson replied:
@wonder-sk I am thinking it will be used like this: https://github.com/nyalldawson/QGIS/blob/tiled_mesh_renderer2/src/core/tiledmesh/qgstiledmeshlayerrenderer.cpp#L644-L680

wonder-sk replied:
I see - but wouldn't it be better to integrate this into getTiles()? Otherwise we would need to implement the same logic in various places where getTiles() may be needed - identify, profile tool, ...

nyalldawson replied:
The issue is that we'd then need two versions of getTiles -- one which is allowed to block during network requests, and the other which somehow allows the event loop to keep processing while requests are handled..

wonder-sk replied:
Ideally, yes. But for the time being, we could live with just one version - the one that blocks - and possibly implement non-blocking variant later. I think users are quite used to the fact that some operations (like identify) can block the GUI :wink: 

Some more arguments for having the logic in getTiles() :slightly_smiling_face: 
- plugin developers using the getTiles() API it may be also difficult to wrap their heads around how to implement the logic... It would be very convenient if they could just use a high-level API to get everything needed and not worry about the details of populating the hierarchy.
- it would be nice if getTiles() would always return the same result irrespective of what bits of hierarchy were already seen in the given index
- getTiles() could optimize network requests - e.g. spawn multiple requests at once and wait for all of them - I assume the logic on the client side would most likely serialize the requests for simplicity

nyalldawson replied:
Ok -- I'm gonna go for the most flexible option and add flags to QgsTiledMeshRequest. By default we'll fetch the hierarchy in a blocking call, but the flags allow for a caller to opt out of this if they are going to handling the hierarchy fetch/refinement themselves.

wonder-sk reviewed (approved):
This approach looks good!

nyalldawson committed:
Add intersects tests for QgsAbstractTiledMeshNodeBoundingVolume

nyalldawson committed:
Move parent tile ID filter to QgsTiledMeshRequest

nyalldawson commented on src/core/tiledmesh/qgstiledmeshindex.h:
```diff
@@ -0,0 +1,239 @@
+     *
+     * May return an empty list if no data satisfies the request.
+     */
+    QStringList getTiles( const QgsTiledMeshRequest &request, const QString &parentId = QString() );
```
Ok, done

nyalldawson committed:
Remove QgsTiledMeshNode from public api

nyalldawson commented on src/core/tiledmesh/qgstiledmeshnode.h:
```diff
@@ -0,0 +1,87 @@
+ *
+ * \since QGIS 3.34
+ */
+class CORE_EXPORT QgsTiledMeshNode
```
Done

nyalldawson committed:
Avoid ambiguity in 'refinement' naming, remove ambiguity about
whether a tile has children available, no children, or not-yet-
fetched children

nyalldawson committed:
Don't assume that json extensions are always present for content
uris

nyalldawson committed:
Add flags to QgsTiledMeshRequest

nyalldawson committed:
By default fetch whole hierarchy when requesting cesium tiles

nyalldawson committed:
Fix seealso links

nyalldawson committed:
Fix qt6 build

wonder-sk reviewed (approved):
Thanks for the updates @nyalldawson - looking great :clap:

nyalldawson committed:
Fix clang tidy warning
